File: android/app/src/main/java/com/dgp/engine/TestVectors.kt
package com.dgp.engine

data class TestVector(
    val seed: String,
    val account: String,
    val name: String,
    val type: String,
    val expected: String
)

object TestVectors {

    val vectors: List<TestVector> = buildList {
        // Basic vectors
        add(TestVector("a", "", "aa", "alnum", "oxToKKV2"))
        add(TestVector("aa", "", "a", "alnum", "g5ELZ6hf"))
        add(TestVector("a", "", "aa", "base58", "zWNoxToK"))
        add(TestVector("a", "", "aa", "alnumlong", "zWNoxToKKV2j"))

        // All types: seed=passwordPASSWORDpassword, account="", name=saltSALT...
        val p = "passwordPASSWORDpassword"
        val s = "saltSALTsaltSALTsaltSALTsaltSALTsalt"
        add(TestVector(p, "", s, "hex", "21934584"))
        add(TestVector(p, "", s, "hexlong", "21934584b8ffde2e"))
        add(TestVector(p, "", s, "alnum", "1Yir2cqz"))
        add(TestVector(p, "", s, "alnumlong", "1Yir2cqzSZNb"))
        add(TestVector(p, "", s, "base58", "1Yir2cqz"))
        add(TestVector(p, "", s, "base58long", "1Yir2cqzSZNb"))
        add(TestVector(p, "", s, "xkcd", "OrdinaryReadyIcePower"))
        add(TestVector(p, "", s, "xkcdlong", "OrdinaryReadyIcePowerVerbStill"))

        // All types: seed=pass, account=word, name=salt
        add(TestVector("pass", "word", "salt", "hex", "842b8a86"))
        add(TestVector("pass", "word", "salt", "hexlong", "842b8a866ef6f789"))
        add(TestVector("pass", "word", "salt", "alnum", "HUgR5fny"))
        add(TestVector("pass", "word", "salt", "alnumlong", "HUgR5fnynuHv"))
        add(TestVector("pass", "word", "salt", "base58", "HUgR5fny"))
        add(TestVector("pass", "word", "salt", "base58long", "HUgR5fnynuHv"))
        add(TestVector("pass", "word", "salt", "xkcd", "StemDialSureHen"))
        add(TestVector("pass", "word", "salt", "xkcdlong", "StemDialSureHenAlbumDonor"))

        val a64 = "A".repeat(64)
        val a65 = "A".repeat(65)
        val b64 = "B".repeat(64)
        val b65 = "B".repeat(65)

        // some_types with no account
        for ((seed, name, hlExp, alExp, xlExp) in listOf(
            Triple5(a64, "salt", "58c3f208e6ddf020", "PARqw6dE", "DivorceAddressHintRuralViciousPeasant"),
            Triple5(a65, "salt", "9d63c79fabd80bc0", "zcXridY3", "ButterTrendPupilTreeSettleWave"),
            Triple5(a64, b64, "2a8702b185b36310", "7Pr3ECx2", "DynamicVacantReadyRifleGuardDestroy"),
            Triple5(a64, b65, "9c414dc4add39239", "7oBwChSZ", "ShrugCargoKeyFollowAcousticPole"),
            Triple5(a65, b64, "2d92337b38d9279e", "qt7hBsti", "DisagreePersonBeginConductGeniusMemory"),
            Triple5(a65, b65, "d253b247c80be550", "entN2FVV", "SquareBehaveDawnToiletUnknownAddict"),
        )) {
            add(TestVector(seed, "", name, "hexlong", hlExp))
            add(TestVector(seed, "", name, "alnum", alExp))
            add(TestVector(seed, "", name, "xkcdlong", xlExp))
        }

        // some_types with account=default
        for ((seed, name, hlExp, alExp, xlExp) in listOf(
            Triple5(a64, "salt", "631817232100b665", "Aqvbp2ce", "AutumnLeopardDonkeyAcquireKeepOrient"),
            Triple5(a65, "salt", "90cee7907f9992c3", "DfWGSPC4", "YoungProudMonitorBabyGaugeCrisp"),
            Triple5(a64, b64, "f7df3aa86e825f90", "yYC6LKxw", "DebateBlessWisdomWillPostSlim"),
            Triple5(a64, b65, "784ee7adec0a3d53", "uCN95mmQ", "GospelDoctorSpiderFrostFleeRequire"),
            Triple5(a65, b64, "a556da50eca68f48", "mTNf5LBU", "TipMatrixWhatRevealDinnerObject"),
            Triple5(a65, b65, "92657b613e27055d", "G8caSAgC", "PioneerSnowTreeToiletWrongOblige"),
        )) {
            add(TestVector(seed, "default", name, "hexlong", hlExp))
            add(TestVector(seed, "default", name, "alnum", alExp))
            add(TestVector(seed, "default", name, "xkcdlong", xlExp))
        }

        // some_types with account=test
        for ((seed, name, hlExp, alExp, xlExp) in listOf(
            Triple5(a64, "salt", "44e62d0c3c6e8c4c", "2aQahrAd", "SideSplitEasyStandKetchupTry"),
            Triple5(a65, "salt", "54bd33e6779b3570", "mTjWwoK8", "EnjoyExtraWheelThrowBeginCompany"),
            Triple5(a64, b64, "7709c029647ebdcb", "v22zGD6C", "VoiceKidDealOffModifyFinal"),
            Triple5(a64, b65, "67a25500a88e1193", "Sawi6Lby", "EmbodyActualAssetBoxChampionAnchor"),
            Triple5(a65, b64, "46ce8f41a45c2fd6", "hDaVCKY9", "IssueArgueSpaceSegmentCasualPlate"),
            Triple5(a65, b65, "c84856500ca1d4b1", "49RT52Ed", "PropertyNetCrossDangerLaughPepper"),
        )) {
            add(TestVector(seed, "test", name, "hexlong", hlExp))
            add(TestVector(seed, "test", name, "alnum", alExp))
            add(TestVector(seed, "test", name, "xkcdlong", xlExp))
        }
    }

    data class Triple5(val a: String, val b: String, val c: String, val d: String, val e: String)

    fun run(engine: DgpEngine): TestResult {
        val lines = mutableListOf<String>()
        var passed = 0
        var failed = 0

        for (tv in vectors) {
            val actual = engine.generate(tv.seed, tv.name, tv.type, tv.account)
            if (actual == tv.expected) {
                passed++
            } else {
                failed++
                val seedDisp = if (tv.seed.length > 16) "${tv.seed.take(8)}...(${tv.seed.length})" else tv.seed
                val nameDisp = if (tv.name.length > 16) "${tv.name.take(8)}...(${tv.name.length})" else tv.name
                lines.add("FAIL [$seedDisp:${tv.account}:$nameDisp:${tv.type}]")
                lines.add("  expected: ${tv.expected}")
                lines.add("  actual:   $actual")
            }
        }

        lines.add(0, "$passed passed, $failed failed out of ${vectors.size} tests")
        return TestResult(passed, failed, lines.joinToString("\n"))
    }

    data class TestResult(val passed: Int, val failed: Int, val output: String) {
        val allPassed get() = failed == 0
    }
}

File: linux/dgp/testvectors.py
from dataclasses import dataclass

from dgp.engine import generate

_A64 = "A" * 64
_A65 = "A" * 65
_B64 = "B" * 64
_B65 = "B" * 65
_P = "passwordPASSWORDpassword"
_S = "saltSALTsaltSALTsaltSALTsaltSALTsalt"


@dataclass(frozen=True)
class TestVector:
    seed: str
    account: str
    name: str
    type: str
    expected: str


VECTORS: list[TestVector] = [
    # 4 basic
    TestVector("a", "", "aa", "alnum", "oxToKKV2"),
    TestVector("aa", "", "a", "alnum", "g5ELZ6hf"),
    TestVector("a", "", "aa", "base58", "zWNoxToK"),
    TestVector("a", "", "aa", "alnumlong", "zWNoxToKKV2j"),
    # all types: seed=passwordPASSWORDpassword, account="", name=saltSALT...
    TestVector(_P, "", _S, "hex", "21934584"),
    TestVector(_P, "", _S, "hexlong", "21934584b8ffde2e"),
    TestVector(_P, "", _S, "alnum", "1Yir2cqz"),
    TestVector(_P, "", _S, "alnumlong", "1Yir2cqzSZNb"),
    TestVector(_P, "", _S, "base58", "1Yir2cqz"),
    TestVector(_P, "", _S, "base58long", "1Yir2cqzSZNb"),
    TestVector(_P, "", _S, "xkcd", "OrdinaryReadyIcePower"),
    TestVector(_P, "", _S, "xkcdlong", "OrdinaryReadyIcePowerVerbStill"),
    # all types: seed=pass, account=word, name=salt
    TestVector("pass", "word", "salt", "hex", "842b8a86"),
    TestVector("pass", "word", "salt", "hexlong", "842b8a866ef6f789"),
    TestVector("pass", "word", "salt", "alnum", "HUgR5fny"),
    TestVector("pass", "word", "salt", "alnumlong", "HUgR5fnynuHv"),
    TestVector("pass", "word", "salt", "base58", "HUgR5fny"),
    TestVector("pass", "word", "salt", "base58long", "HUgR5fnynuHv"),
    TestVector("pass", "word", "salt", "xkcd", "StemDialSureHen"),
    TestVector("pass", "word", "salt", "xkcdlong", "StemDialSureHenAlbumDonor"),
    # 6×3 with account=""
    TestVector(_A64, "", "salt", "hexlong", "58c3f208e6ddf020"),
    TestVector(_A64, "", "salt", "alnum", "PARqw6dE"),
    TestVector(_A64, "", "salt", "xkcdlong", "DivorceAddressHintRuralViciousPeasant"),
    TestVector(_A65, "", "salt", "hexlong", "9d63c79fabd80bc0"),
    TestVector(_A65, "", "salt", "alnum", "zcXridY3"),
    TestVector(_A65, "", "salt", "xkcdlong", "ButterTrendPupilTreeSettleWave"),
    TestVector(_A64, "", _B64, "hexlong", "2a8702b185b36310"),
    TestVector(_A64, "", _B64, "alnum", "7Pr3ECx2"),
    TestVector(_A64, "", _B64, "xkcdlong", "DynamicVacantReadyRifleGuardDestroy"),
    TestVector(_A64, "", _B65, "hexlong", "9c414dc4add39239"),
    TestVector(_A64, "", _B65, "alnum", "7oBwChSZ"),
    TestVector(_A64, "", _B65, "xkcdlong", "ShrugCargoKeyFollowAcousticPole"),
    TestVector(_A65, "", _B64, "hexlong", "2d92337b38d9279e"),
    TestVector(_A65, "", _B64, "alnum", "qt7hBsti"),
    TestVector(_A65, "", _B64, "xkcdlong", "DisagreePersonBeginConductGeniusMemory"),
    TestVector(_A65, "", _B65, "hexlong", "d253b247c80be550"),
    TestVector(_A65, "", _B65, "alnum", "entN2FVV"),
    TestVector(_A65, "", _B65, "xkcdlong", "SquareBehaveDawnToiletUnknownAddict"),
    # 6×3 with account="default"
    TestVector(_A64, "default", "salt", "hexlong", "631817232100b665"),
    TestVector(_A64, "default", "salt", "alnum", "Aqvbp2ce"),
    TestVector(_A64, "default", "salt", "xkcdlong", "AutumnLeopardDonkeyAcquireKeepOrient"),
    TestVector(_A65, "default", "salt", "hexlong", "90cee7907f9992c3"),
    TestVector(_A65, "default", "salt", "alnum", "DfWGSPC4"),
    TestVector(_A65, "default", "salt", "xkcdlong", "YoungProudMonitorBabyGaugeCrisp"),
    TestVector(_A64, "default", _B64, "hexlong", "f7df3aa86e825f90"),
    TestVector(_A64, "default", _B64, "alnum", "yYC6LKxw"),
    TestVector(_A64, "default", _B64, "xkcdlong", "DebateBlessWisdomWillPostSlim"),
    TestVector(_A64, "default", _B65, "hexlong", "784ee7adec0a3d53"),
    TestVector(_A64, "default", _B65, "alnum", "uCN95mmQ"),
    TestVector(_A64, "default", _B65, "xkcdlong", "GospelDoctorSpiderFrostFleeRequire"),
    TestVector(_A65, "default", _B64, "hexlong", "a556da50eca68f48"),
    TestVector(_A65, "default", _B64, "alnum", "mTNf5LBU"),
    TestVector(_A65, "default", _B64, "xkcdlong", "TipMatrixWhatRevealDinnerObject"),
    TestVector(_A65, "default", _B65, "hexlong", "92657b613e27055d"),
    TestVector(_A65, "default", _B65, "alnum", "G8caSAgC"),
    TestVector(_A65, "default", _B65, "xkcdlong", "PioneerSnowTreeToiletWrongOblige"),
    # 6×3 with account="test"
    TestVector(_A64, "test", "salt", "hexlong", "44e62d0c3c6e8c4c"),
    TestVector(_A64, "test", "salt", "alnum", "2aQahrAd"),
    TestVector(_A64, "test", "salt", "xkcdlong", "SideSplitEasyStandKetchupTry"),
    TestVector(_A65, "test", "salt", "hexlong", "54bd33e6779b3570"),
    TestVector(_A65, "test", "salt", "alnum", "mTjWwoK8"),
    TestVector(_A65, "test", "salt", "xkcdlong", "EnjoyExtraWheelThrowBeginCompany"),
    TestVector(_A64, "test", _B64, "hexlong", "7709c029647ebdcb"),
    TestVector(_A64, "test", _B64, "alnum", "v22zGD6C"),
    TestVector(_A64, "test", _B64, "xkcdlong", "VoiceKidDealOffModifyFinal"),
    TestVector(_A64, "test", _B65, "hexlong", "67a25500a88e1193"),
    TestVector(_A64, "test", _B65, "alnum", "Sawi6Lby"),
    TestVector(_A64, "test", _B65, "xkcdlong", "EmbodyActualAssetBoxChampionAnchor"),
    TestVector(_A65, "test", _B64, "hexlong", "46ce8f41a45c2fd6"),
    TestVector(_A65, "test", _B64, "alnum", "hDaVCKY9"),
    TestVector(_A65, "test", _B64, "xkcdlong", "IssueArgueSpaceSegmentCasualPlate"),
    TestVector(_A65, "test", _B65, "hexlong", "c84856500ca1d4b1"),
    TestVector(_A65, "test", _B65, "alnum", "49RT52Ed"),
    TestVector(_A65, "test", _B65, "xkcdlong", "PropertyNetCrossDangerLaughPepper"),
]


@dataclass(frozen=True)
class SingleResult:
    label: str
    expected: str
    actual: str
    passed: bool


def run_one(idx: int) -> SingleResult:
    tv = VECTORS[idx]
    actual = generate(tv.seed, tv.name, tv.type, tv.account)
    passed = actual == tv.expected
    seed_disp = f"{tv.seed[:4]}...({len(tv.seed)})" if len(tv.seed) > 16 else tv.seed
    name_disp = f"{tv.name[:4]}...({len(tv.name)})" if len(tv.name) > 16 else tv.name
    label = f"{seed_disp}:{tv.account}:{name_disp}:{tv.type}"
    return SingleResult(label, tv.expected, actual, passed)


def run_all() -> list[SingleResult]:
    return [run_one(i) for i in range(len(VECTORS))]
